Best Practices with Git

darrenwee opened the pull request:
This is the first draft of my book chapter on git best practices. I've highlighted some important pointers that I've learnt about (most of them the hard way) since I started using git.

Can you take a look @vivekscl :grinning: 

### [view the rendered version here](https://github.com/darrenwee/learningresources/blob/git-best-practices/contents/revisionControl/bestPracticesGit.md)

darrenwee committed:
Add first draft of Best Practices with Git.

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+## Write Good Commit Messages
+Good commit messages can help:
+- reviewers to understand the high-level changes in your pull request/patch
+- others to understand your reasoning behind changes made by that commit,
```
... reasoning behind **the** changes made...

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+- [torvalds/linux](https://github.com/torvalds/linux/commits/master)
+
+### Set Up Your Editor for Commit Messages
+1. To use your editor of choice for `git`-related functionality, e.g. `vim`, do one of either
```
Maybe you could explicitly say to type the following in a terminal like `do one of either in your terminal`?

vivekscl replied:
Since point 2 has a `:` at the end, point 1 should have a `:` at the end.

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
````diff
@@ -0,0 +1,259 @@
+### Stashing
+If you need to switch between branches while in the middle of developing a commit, you can use the `git stash` command instead of committing your half-done code.
+
+```bash
````
You stated `bash` here but there doesn't seem to be any syntax highlighting, so is this needed? Furthermore, the `bash` keyword wasn't used in the above section about `Set Up Your Editor for Commit Messages`.

darrenwee replied:
The syntax highlighting is for the bash comments so that it doesn't look the same as the git commands.

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+- Typically a hyphen or asterisk is used for the bullet, followed by a
+  single space, with blank lines in between, but conventions vary here
+
+- Use a hanging indent
```
Not sure what a hanging indent is though :sweat_smile: . Maybe you should explain it?

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
````diff
@@ -0,0 +1,259 @@
+git stash list
+```
+
+Stashes are a purely local construct and cannot be pushed to a remote repository.
````
Maybe add this as a note instead?

darrenwee replied:
I left it as-is because I'm presenting it as an alternative to committing WIP code, so it should be in the body.

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+    - changing representation of data to a different format
+
+Each _logical change in code_ should translate to exactly one commit, nothing more or less. Doing so allows you to:
+- revert a particular logical change with no/less side effects
```
Does `with little to no side effects` sound better?

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+Each _logical change in code_ should translate to exactly one commit, nothing more or less. Doing so allows you to:
+- revert a particular logical change with no/less side effects
+- easily identify bad commits that caused change in behavior
+- collaborate with others, e.g. by cherry-picking a single logical change instead of finding a bit of this commit and a bit of that commit
```
Should it be, `collaborate with others easily` instead? Since you're emphasizing the benefits, just stating `collaborate with others` doesn't seem as beneficial.

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+---
+
+## Respect Published History
+Always avoid rewriting the public, published history unless you have a very good reason, you are very sure of what you are doing, or you will not affect anyone else.
```
Maybe you could say `Always avoid rewriting the public history` or `Always avoid rewriting the published history` instead? Putting both terms in there seems redundant.

vivekscl replied:
If there is an `Exceptions` section below, is this part needed? 
`unless you have a very good reason, you are very sure of what you are doing, or you will not affect anyone else`

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+- [Git Basics - Working with Remotes](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes)
+
+### Rebase versus Merging
+| You should... | when ... |
```
To standardize the headers, `when ...` should have the `w` capitalized and the space removed between `when` and `...` like so:
**W**hen**...**

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+
+It is generally considered good practice to rebase your feature branch onto whatever branch you're trying to patch _before_ you make the pull request, resolving any conflicts that arise. This:
+
+- [keeps the history clear and linear](http://www.bitsnbites.eu/a-tidy-linear-git-history/).
```
An unnecessary period at the end of this line.

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+## Hide the Sausage Making
+_Sausage making_ refers to the process by which code is incrementally worked on, where a series of commits (like links in a sausage) make up a branch. It is often desirable to _hide the sausage making_ where the commit history is cleaned up so that it looks neater and is easier to follow.
+
+When working on a `feature`/`fix` branch, you may
```
Missing `:` at the end

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+- performing an interactive rebase, i.e. `git rebase -i`
+- patch-wise reset and stage, i.e. `git reset -p` and `git add -p`
+
+Ensure that you do this _before_
```
Missing `:` at the end

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+
+Stashes are a purely local construct and cannot be pushed to a remote repository.
+
+Read more: [git stash - Saving Changes | Atlassian Git Tutorial](https://www.atlassian.com/git/tutorials/git-stash)
```
Maybe there should be a newline after the `:` and bullet point like the other sections?
![deepinscreenshot_select-area_20180305162937](https://user-images.githubusercontent.com/17037188/36964749-9a2aafc6-2092-11e8-9a20-516cb20d291a.png)

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+Changes to your own branches that is used by others should obey always-functioning-commits rule to minimize surprise. This is especially important if you expect your branch to be cherry-picked by another collaborator because they require a specific bit of code that you wrote.
+
+### Stashing
+If you need to switch between branches while in the middle of developing a commit, you can use the `git stash` command instead of committing your half-done code.
```
Maybe what stash does should be explicitly stated?
you can use the `git stash` command **to save your code** instead of...

vivekscl commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+If you have made several overlapping changes on your working directory (e.g. forgot to commit, etc), you can always [perform a patch-wise stage using `git add -p`](https://stackoverflow.com/questions/1085162/commit-only-part-of-a-file-in-git).
+
+Read more:
+- [programming practices - When to commit code?](https://softwareengineering.stackexchange.com/questions/83837/when-to-commit-code)
```
`P`rogramming practices

darrenwee replied:
I will not implement this change as it's the original title of the link, so I want to preserve it as-is.

vivekscl reviewed (changes requested):
Several standardization issues and suggestions. 
Content-wise, it's really informative and well structured :+1:

Zhiyuan-Amos commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+    - makes backtracking easier
+    - easy to follow history
+    - reverting/rolling back is much simpler
+    - you can use `git bisect` to find regressions easily
```
You can use `git bisect` even without rebasing to master right? Cos you can git bisect on `master`, and once you find out which `merge` is buggy, then `git bisect` the `merge/PR`.

Zhiyuan-Amos commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+    - easy to follow history
+    - reverting/rolling back is much simpler
+    - you can use `git bisect` to find regressions easily
+- ensures your changes are compatible with the head of the branch you're patching
```
This is already mentioned above

> It is generally considered good practice to rebase your feature branch onto whatever branch you're trying to patch _before_ you make the pull request, **resolving any conflicts that arise.**

darrenwee replied:
This point (line 227) is supposed to be about the benefit of rebasing onto the base branch _before_ making a PR. Resolving conflicts during the rebase process guarantees that your branch is compatible with the base branch that you're patching, so it's not quite the same point I'm trying to put across. One is cause and the other is effect. What do you think?

Zhiyuan-Amos replied:
Ah ok I see! :)

Zhiyuan-Amos commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+---
+
+## Keep Up To Date
+### Use Remotes Effectively
```
The title can be better phrased because this seems more of like `How to add a remote` rather than `use remotes effectively`.

Zhiyuan-Amos commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+
+---
+
+## Respect Published History
```
This segment seems a bit short compared to the rest, and also the idea of `force pushing` to rewrite master branch aren't mentioned in the links as well.

darrenwee replied:
I beefed this section up and added a bit on force pushing.

Zhiyuan-Amos commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+
+---
+
+## One Logical Change to One Commit
```
You can give examples of good commits. Perhaps... Get from `AddressBook`? :P

darrenwee replied:
Haha sure, thanks!!

Zhiyuan-Amos commented on contents/revisionControl/bestPracticesGit.md:
````diff
@@ -0,0 +1,259 @@
+git config --global core.editor "vim" # or you can do the following
+export GIT_EDITOR=vim # add to your .bashrc
+```
+2. Set your editor to wrap after 72 characters. In `vim`, you can do this by adding this to your `.vimrc`:
````
You can mention why we should set the editor to wrap after 72 characters.

tshradheya commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,259 @@
+the body entirely); tools like rebase can get confused if you run the
+two together.
+
+Write your commit message in the imperative: "Fix bug" and not "Fixed bug"
```
One good thing I have read in many articles related to writing git commit messages is to make sure the message makes sense when read as follows:

>If applied, this commit will `your commit message here`

Maybe can add this here?

darrenwee replied:
I added it outside of the commit message code block so that it stands out more.

darrenwee committed:
Fix formatting and phrasing based on Viv's review.

darrenwee committed:
Revise according to Zhiyuan/Shradheya's reviews.

- Beef up section on Respecting Published History.
- Revise section on Writing a Good Commit Message.
- Add examples of good commit messages from addressbook.
- Clarify use of git bisect.

darrenwee committed:
Fix formatting for commit litmus test.

darrenwee commented:
@damithc I'm ready for a review!

Zhiyuan-Amos commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,311 @@
+- reviewers to understand the high-level changes in your pull request/patch
+- others to understand your reasoning behind the changes made by that commit,
+    - while reviewing your code,
+    - while figuring out why a piece of code that is five years old is that way,
```
the comma here is kinda weird.

darrenwee replied:
I will revise the structure to clean this part up and remove the redundant `while` and commas.

Zhiyuan-Amos reviewed (commented):
Sorry, don't have time to give a thorough look, so I won't give an explicit approve / request changes :<

darrenwee committed:
Restructure Write Good Commit Messages section.

Reword some stuff for clarity, and add more pros to writing good commit
messages!

damithc commented:
Had a quick look. No major objections. Get approval from a peer first?

vivekscl reviewed (approved):
Looks good to me!

darrenwee commented:
@damithc Sorry about that, it's approved now :smile:

damithc commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,313 @@
+
+---
+
+## One Logical Change to One Commit
```
One Logical Change Per Commit?

damithc commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,313 @@
+## One Logical Change to One Commit
+Commits are the building blocks of a codebase; each building block should contribute exactly one useful thing, like:
+- adding a new function or piece of data
+- fixing one or more bugs
```
Fixing multiple bugs in one commit contradicts 'one logical change per commit'?

damithc commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,313 @@
+Source: [A Note About Git Commit Messages by Tim Pope.](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)
+
+As a litmus test, you can try to read your commit message summary in the following manner:
+> If applied, this commit will `your commit message here`
```
`your commit message summary here`? 

Also, can give an example

damithc commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,313 @@
+
+Write your commit message in the imperative: "Fix bug" and not "Fixed bug"
+or "Fixes bug."  This convention matches up with commit messages generated
+by commands like git merge and git revert.
```
This may be useful https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

darrenwee replied:
I decided to scrap the old format for describing a good commit message and I took the list from that link and reorganized it a bit.

damithc commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,313 @@
+        - [Stashing](#stashing)
+    - [One Logical Change to One Commit](#one-logical-change-to-one-commit)
+    - [Hide the Sausage Making](#hide-the-sausage-making)
+        - [How to Hide the Sausage](#how-to-hide-the-sausage)
```
`Hide the Sausage` or `Hide the Sausage Making`?

damithc commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,313 @@
+- makes reviewing/testing easier by [not including irrelevant code by merging](https://lwn.net/Articles/328436/)
+
+Read more:
+- [A Tidy, Linear Git History](http://www.bitsnbites.eu/a-tidy-linear-git-history/) _this is an excellent article which has formed my opinion on rebase vs. merge_
```
It is good to give an overview of each resource like you have done here. Try to do it more. But may be not use italics?

darrenwee replied:
I wasn't sure what was the best way to do this, so I took out the italics and uses a hyphen to delimit the start of the overview. I only give an overview for the more important/interesting resources, though. Some like the StackOverflow threads are self-explanatory.

damithc reviewed (changes requested):
Some suggestions

darrenwee committed:
Revise draft according to Damith's comments.

darrenwee commented:
@damithc I made some structural changes to the section on commit messages; otherwise everything else are minor edits as per your comments :smile:

damithc reviewed (changes requested):
Looks good. As you have given very detailed guidelines for  commit message writing, it would be good to give one or two concrete examples (some people prefer to learn from examples). You can reuse the examples from oss-generic page (remember to cite the source).

rachx commented on contents/revisionControl/bestPracticesGit.md:
```diff
@@ -0,0 +1,312 @@
+        - [Set Up Your Editor for Commit Messages](#set-up-your-editor-for-commit-messages)
+    - [Always Commit Functional Code](#always-commit-functional-code)
+        - [Stashing](#stashing)
+    - [One Logical Change to One Commit](#one-logical-change-to-one-commit)
```
Link broken

darrenwee committed:
Add good commit message examples

Adds two (good) commit message examples from the Linux kernel project
and the addressbook-level4 project.

darrenwee committed:
Fix broken reference to #one-logical-change-per-commit

Forgot to change this when I renamed the section.

darrenwee committed:
Add link to oss-generic commit message examples.

darrenwee commented:
I've fixed the broken reference and added some examples of good commit messages! I was tempted to add some more cool git tools like `git bisect` or the `reflog` but I think that can be left for a different chapter (or someone else to write).

damithc reviewed (approved):
LGTM
